shiny/ui/_toast.py

```diff
@@ -0,0 +1,554 @@
+        """Convert to HTML Tag object."""
+        # Danger type requires more assertive accessibility attributes
+        if self.type == "danger":
+            role = "alert"
+            aria_live = "assertive"
+        else:
+            role = "status"
+            aria_live = "polite"
+
+        classes = ["toast"]
+        if self.type:
+            classes.append(f"text-bg-{self.type}")
+
+        contents: list[TagChild] = []
+
+        close_button = tags.button(
+            type="button",
+            class_="btn-close",
+            data_bs_dismiss="toast",
+            aria_label="Close",
+        )
+
+        if self.header is not None:
+            if isinstance(self.header, ToastHeader) or is_tag_node(self.header):
+                header = self.header
+            else:
+                header = toast_header(title=self.header)
+
+            if isinstance(header, ToastHeader):
+                header_tag = header.tagify(
+                    close_button=close_button if self.closable else None
+                )
+            else:
+                header_tag = div(
+                    header,
+                    close_button if self.closable else None,
+                    class_="toast-header",
+                )
+
+            contents.append(header_tag)
+
+        # Close button placement: header if present, otherwise body
+        body_has_close_btn = self.header is None and self.closable
+
+        if not body_has_close_btn and self.icon is None:
+            body_tag = div({"class": "toast-body"}, self.body)
+        else:
+            body_contents: list[TagChild] = []
+
+            if self.icon is not None:
+                body_contents.append(tags.span(self.icon, class_="toast-body-icon"))
+
+            body_contents.append(
+                div({"class": "toast-body-content flex-grow-1"}, self.body)
+            )
+
+            if body_has_close_btn:
+                body_contents.append(close_button)
+
+            body_tag = div({"class": "toast-body d-flex gap-2"}, *body_contents)
+
+        contents.append(body_tag)
+
+        return div(
+            {
+                "class": " ".join(classes),
+                "id": id or self.id or _toast_random_id(),
+            },
+            *contents,
+            components_dependencies(),
+            role=role,
+            aria_live=aria_live,
+            aria_atomic="true",
+            data_bs_autohide="true" if self.autohide else "false",
+            **self.attribs,
+        )
+
+
+@add_example()
+def toast(
+    *args: TagChild | TagAttrs,
+    header: Optional[str | ToastHeader | TagNode] = None,
+    icon: Optional[TagNode] = None,
+    id: Optional[str] = None,
+    type: Optional[
+        Literal[
+            "primary",
+            "secondary",
+            "success",
+            "info",
+            "warning",
+            "danger",
+            "error",
+            "light",
+            "dark",
+        ]
```

Comment:
Could be worth saving this in a variable and using it to narrow `type` in other contexts, like `_normalize_toast_type()`?

Reply:
Good idea, done!